### PR TITLE
refactor: improve zero address

### PR DESCRIPTION
**Pull request by septs**
## Description

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [x] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/background-script/EthereumServices/rpc/index.ts [hunk 1]
```diff
@@ -132,7 +133,7 @@ export async function getSendTransactionComputedPayload(payload: JsonRpcPayload)
         }
 
         // contract deployment
-        if (isSameAddress(to, ZERO_ADDRESS)) {
+        if (isZeroAddress(to)) {
```
Comment (guanbinrui) [inline packages/mask/src/extension/background-script/EthereumServices/rpc/index.ts:136]: No need to remove the importing of `ZERO_ADDRESS`.

File: packages/mask/src/plugins/MaskBox/hooks/useContext.ts [hunk 2]
```diff
@@ -65,7 +67,7 @@ function useContext(initialState?: { boxId: string }) {
     const { value: paymentTokens = [] } = useFungibleTokensDetailed(
         maskBoxStatus?.payment?.map(([address]) => {
             return {
-                type: isSameAddress(address, ZERO_ADDRESS) ? EthereumTokenType.Native : EthereumTokenType.ERC20,
+                type: isNative(address) ? EthereumTokenType.Native : EthereumTokenType.ERC20,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/MaskBox/hooks/useContext.ts:70]: Rename it to `isNativeTokenAddress`.

File: packages/web3-shared/evm/utils/address.ts [hunk 3]
```diff
@@ -1,24 +1,32 @@
...
+import { toArray } from 'lodash-unified'
 import { EthereumAddress } from 'wallet.ts'
-import { getRedPacketConstants, getTokenConstants } from '../constants'
+import { getRedPacketConstants, getTokenConstants, ZERO_ADDRESS } from '../constants'
+import { ChainId } from '../types'
 
-export function isSameAddress(addrA: string = '', addrB: string = '') {
-    if (!addrA || !addrB) return false
-    return addrA.toLowerCase() === addrB.toLowerCase()
+export function isSameAddress(source = '', target = '') {
```
Comment (guanbinrui) [inline packages/web3-shared/evm/utils/address.ts:7]: They are not the source vs. target relation.

File: packages/mask/src/plugins/MaskBox/hooks/useContext.ts (hunk 2 above)
Comment (septs) [reply packages/mask/src/plugins/MaskBox/hooks/useContext.ts:70]: CleanShot 2021-12-27 at 13 54 57@2x

File: packages/web3-shared/evm/utils/address.ts [hunk 4]
```diff
@@ -1,24 +1,32 @@
...
         }
         throw new Error('Unsupported `target` address format')
     }
 }
 
-export const isNative = currySameAddress(getTokenConstants().NATIVE_TOKEN_ADDRESS)
+export const isZeroAddress = currySameAddress(ZERO_ADDRESS)
+
+export const isNative = currySameAddress(
+    getEnumAsArray(ChainId).map(({ value }) => getTokenConstants(value).NATIVE_TOKEN_ADDRESS!),
```
Comment (guanbinrui) [inline packages/web3-shared/evm/utils/address.ts:28]:
```suggestion
    getEnumAsArray(ChainId).map(({ value }) => getTokenConstants(value).NATIVE_TOKEN_ADDRESS),
```

File: packages/mask/src/extension/background-script/EthereumServices/rpc/index.ts (hunk 1 above)
Comment (septs) [reply packages/mask/src/extension/background-script/EthereumServices/rpc/index.ts:136]: https://github.com/DimensionDev/Maskbook/blob/7c00da4c23bfb576a090b8a6b35137cbc2c54208/packages/mask/src/extension/background-script/EthereumServices/rpc/index.ts#L32-L37

File: packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketConfirmDialog.tsx [hunk 5]
```diff
@@ -93,7 +92,7 @@ export function RedPacketConfirmDialog(props: ConfirmRedPacketFormProps) {
             <Grid item xs={6}>
                 <Typography variant="body1" color="textPrimary" align="right" className={classes.token}>
                     <span>{settings?.token?.symbol}</span>
-                    {isNative(settings?.token?.address!) ? null : (
+                    {isNativeTokenAddress(settings?.token?.address!) ? null : (
```
Comment (guanbinrui) [inline packages/mask/src/plugins/RedPacket/SNSAdaptor/RedPacketConfirmDialog.tsx:95]:
```suggestion
                    {isNativeTokenAddress(settings?.token?.address) ? null : (
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**